docs/debugging_guide.md

```diff
@@ -0,0 +1,362 @@
+
+## Diagnose Crashes
+
+You can use a debugger to assess the cause of a crash or runtime error, including:
+
+- Failed debugging assertions
+- Undefined behavior detected by sanitizer tools
+- Segmentation faults
+- Unhandled exceptions
+- And more!
+
+Basically, if your code is doing something bad:
+1. Run through the debugger. No breakpoints needed.
+2. The debugger pauses when the error occurs.
+3. You look at local variables, the call stack, etc. to diagnose the issue
+
+Here's several examples:
+
+### Crash in _My_ Code
+
+I've just finished part of the Euchre project and want to run my own `Player_tests.exe` via the terminal:
+
+```console
+$ ./Player_tests.exe
+Running test: test_get_name
+Segmentation fault
+```
+
+My code crashed with a segmentation fault, but I don't know much else.
+
+Let's run in a debugger and let it crash there. No need to set any breakpoints.
+
+<img src="images/debug_crash_0.png" width="800px" />
+
+From the above, we can quickly see the line where the segfault occurs. We also observe that `p1->get_name()` crashes because `p1` is `0x0` (a null pointer), as seen in the variables panel. Ah ha - we just mistyped `p1` on this line instead of `p2`.
+
+Sometimes you might also need to explore [the call stack](#the-call-stack) to determine the cause of the crash. 
+
+### Crash in _Library/System_ Code
+
+Sometimes a crash occurs in library code that your program uses, but that you didn't write.
+
+For example, the debugger might show you a crash in code from the standard library, which is often very difficult to read:
+
+<img src="images/debug_crash_stl.png" width="600px" />
+
+Or, the debugger might choose to show you the compiled assembly code where the crash occurred:
+
+<img src="images/debug_crash_asm.png" width="600px" />
+
+There are also some implicitly defined functions that don't literally appear in your source code (e.g. a built-in copy constructor or assignment operator). For a crash in any of these functions, the debugger may just show you the first line of the class:
+
+<img src="images/debug_crash_implicit.png" width="600px" />
+
+**Tracing Back to _Your_ Code**  
+To diagnose crashes where the debugger initially shows you library code, you need to find the nearest part of _your_ code that ultimately called the library functions.
+
+Let's consider again the case of a crash in an implicitly-defined function. The debugger first just shows us the top line of the class definition, since there's no code to show for implicitly-defined functions.
+
+<img src="images/debug_crash_implicit_full.png" width="800px" />
+
+In this case, check the call stack. First, observe that the segfault did occur in an implicitly-defined function, the built-in `Card::operator=` assignment operator for copying a `Card`.
```

Comment:
Opportunity to link back to the above section!

Reply:
Done. c757b03